### PR TITLE
Adds MeeplePlacement

### PR DESCRIPTION
**Pull request by LeeARichardson**
Puts tile placement data in a new object within Placement, alongside MeeplePlacement.
As such, a lot of changes are fixes for this change.

### CODE REVIEW

**Comment by nadinengland** [comment]
This all look good, my only thoughts are around [`Placement#AddMeeple` method](https://github.com/overengineering/sharpasonne/pull/13/files#diff-0f642e9db7fd1c3376affe56d549407dR17)

Technically, it is more of a `SetOrUpdateMeeple` as it doesn't deal with state change, just sets the property. How do we feel about having two classes to represent the states of with meeple and without. Then each class would have the respective `AddMeeple` or `RemoveMeeple` methods etc?

File: Sharpasonne.Models/Meeple.cs [hunk 1]
```diff
@@ -0,0 +1,12 @@
+namespace Sharpasonne.Models
+{
+    public class Meeple
+    {
+        public Meeple(int player)
```
Comment (StefanoChiodino) [inline Sharpasonne.Models/Meeple.cs]: -> playerIndex/Id

File: Sharpasonne.Models/Placement.cs [hunk 2]
```diff
@@ -1,22 +1,28 @@
...
+using JetBrains.Annotations;
+using Optional;
 
 namespace Sharpasonne.Models
 {
     public class Placement
     {
-        public Tile Tile { get; }
-        public Orientation Orientation { get; }
+        [NotNull] public TilePlacement           TilePlacement   { get; }
```
Comment (StefanoChiodino) [inline Sharpasonne.Models/Placement.cs]: no horiz allinment, it's going to fuck up git history.

File: Sharpasonne/GameActions/EngineState.cs [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+{
+    public class EngineState : IEngine
+    {
+        public EngineState(Board board, IImmutableDictionary<Type, IImmutableList<IRule<IGameAction>>> rules)
+        {
+            Board = board;
+            Rules = rules;
+        }
+
+        public Board                                                          Board { get; }
```
Comment (StefanoChiodino) [inline Sharpasonne/GameActions/EngineState.cs]: no hor all

File: Sharpasonne.Models/Placement.cs (hunk 2 above)
Comment (LeeARichardson) [reply Sharpasonne.Models/Placement.cs]: Oh I'll never stop.

File: Sharpasonne.Models/Meeple.cs (hunk 1 above)
Comment (LeeARichardson) [reply Sharpasonne.Models/Meeple.cs]: Got an issue in for that.

**Pull request merged**